File: 0323NumberofConnectedComponentsinanUndirectedGraph/NumberofConnectedComponentsinanUndirectedGraph.cpp

#include <iostream>
#include <vector>
using std::vector;
#include <unordered_set>
using std::unordered_set;

class DSU {
    int parts;
    vector<int> parents;
    vector<int> heights;


public:
    DSU(int n): parts(n), parents(vector<int>(n)), heights(vector<int>(n, 1)) {
        for (int i = 0; i < n; ++i) {
            parents[i] = i;
        }
    }

    void union_(int a, int b) {
        int ra = this->find(a);
        int rb = this->find(b);
        if (ra != rb) {
            if (this->heights[ra] < this->heights[rb]) {
                this->parents[ra] = rb;
            } else if (this->heights[ra] > this->heights[rb]) {
                this->parents[rb] = ra;
            } else {
                this->parents[rb] = ra;
                this->heights[ra]++;
            }
            this->parts--;
        }
    }

    int getParts() const {
        return this->parts;
    }

private:
    int find(int node) {
        if (this->parents[node] != node) {
            this->parents[node] = find(this->parents[node]);
        }
        return this->parents[node];
    }
};

class NumberOfConnectedComponentsInAnUndirectedGraph {
public:
    int countComponents1(int n, vector<vector<int> >& edges) {
        // adj mat
        vector<vector<bool> > adj(n, vector<bool>(n, false));
        for (size_t i = 0; i < edges.size(); ++i) {
            adj[edges[i][0]][edges[i][1]] = true;
            adj[edges[i][1]][edges[i][0]] = true;
        }
        // visited
        unordered_set<int> visited;
        int ans = 0;
        for (size_t i = 0; i < (size_t)n; ++i) {
            if (visited.find(i) == visited.end()) {
                // one node that is not connected to any previous component
                ans++;
                // mark all nodes connected to this one visited
                this->dfs(adj, visited, (int)i);
            }
        }
        return ans;
    }

    int countComponents3(int n, vector<vector<int> >& edges) {
        DSU dsu(n);
        for (size_t i = 0; i < edges.size(); ++i) {
            dsu.union_(edges[i][0], edges[i][1]);
        }
        return dsu.getParts();
    }

private:
    void dfs(
        vector<vector<bool> >& adj, unordered_set<int>& visited, int node
    ) {
        visited.insert(node);
        for (size_t i = 0; i < adj.size(); ++i) {
            if (adj[node][i] && visited.find((int)i) == visited.end()) {
                this->dfs(adj, visited, (int)i);
            }
        }
    }
};


int main() {
    NumberOfConnectedComponentsInAnUndirectedGraph test;
    vector<vector<int> > edges1;
    vector<int> e1;
    e1.push_back(0);
    e1.push_back(1);
    vector<int> e2;
    e2.push_back(0);
    e2.push_back(2);
    edges1.push_back(e1);
    edges1.push_back(e2);
    // int result2 = solution.countComponents1(n, edges);
    std::cout << test.countComponents1(3, edges1) << std::endl;
    // std::cout << "Number of connected components (DSU): " << result2 << std::endl;
    return EXIT_SUCCESS;
}

File: 0323NumberofConnectedComponentsinanUndirectedGraph/NumberofConnectedComponentsinanUndirectedGraph.py
# -*- coding: utf-8 -*-

from typing import List, Set

class NumberofConnectedComponentsinanUndirectedGraph:
    def countComponents1(self, n: "int", edges: "List[List[int]]") -> "int":
        ## create adj mat
        adj = [[False for _ in range(n)] for _ in range(n)]
        for [a, b] in edges:
            adj[a][b] = True
            adj[b][a] = True
        visited = set()
        ans = 0
        for i in range(n):
            if i not in visited: ## found a new component
                ans += 1 ## increment ans
                ## mark all connected nodes as visted
                self.__dfs(adj, visited, i)
        return ans

    def __dfs(
        self, adj: "List[List[bool]]", visited: "Set[int]", node : "int"
    ) -> "None":
        visited.add(node)
        for i in range(len(adj)):
            if adj[node][i] and i not in visited:
                visited.add(i)
                self.__dfs(adj,visited, i)

    def countComponents3(self, n: "int", edges: "List[List[int]]") -> "int":
        dsu = DSU(n)
        for [a, b] in edges:
            dsu.union(a, b)
        return dsu.getParts()


class DSU:
    def __init__(self, n: "int"):
        self.__parents = [i for i in range(n)]
        self.__heights = [1 for _ in range(n)]
        self.__parts = n

    def union(self, a: "int", b: "int") -> "None":
        ra = self.__find(a)
        rb = self.__find(b)
        if ra != rb:
            if self.__heights[ra] < self.__heights[rb]:
                self.__parents[ra] = rb
            elif self.__heights[ra] > self.__heights[rb]:
                self.__parents[rb] = ra
            else:
                self.__parents[rb] = ra
                self.__heights[ra] += 1
            self.__parts -= 1

    def __find(self, n: "int") -> "int":
        """
        Find the root of a given node
        """
        if self.__parents[n] != n:
            self.__parents[n] = self.__find(self.__parents[n])
        return self.__parents[n]

    def getParts(self) -> "int":
        return self.__parts


def main():
    test = NumberofConnectedComponentsinanUndirectedGraph()
    print(test.countComponents3(3, [[0, 1], [0, 2]]))
    print(test.countComponents3(5, [[0, 1], [1, 2], [3, 4]]))
    print(test.countComponents3(6, [[0, 1], [1, 2], [2, 3], [4, 5]]))

if __name__ == "__main__":
    main()
